### PR TITLE
LP-1827 Show specialisation tag on course card for courses which are part of specialisation

### PR DESCRIPTION
**Pull request by monofal**
### Description

[LP-1827](https://philanthropyu.atlassian.net/browse/LP-1827)

[edx-theme PR](https://github.com/philanthropy-u/philu-edx-theme/pull/1248)

Show specialization tag on catalog page with courses that are part of specialization/programs.

### CODE REVIEW

File: openedx/features/course_card/views.py [hunk 1]
```diff
@@ -48,13 +50,25 @@ def get_course_cards(request):
         if course.invitation_only and not CourseEnrollment.is_enrolled(request.user, course.id):
             continue
 
+        programs = get_programs(request.site)
```
[the remark this answers is not in the document]
Comment (monofal) [reply openedx/features/course_card/views.py]: done

File: openedx/features/course_card/views.py [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
-from openedx.features.course_card.models import CourseCard
-from django.views.decorators.csrf import csrf_exempt
 from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from openedx.features.course_card.models import CourseCard
+from philu_overrides.helpers import get_user_current_enrolled_class
 from student.models import CourseEnrollment
+
+from ...core.djangoapps.catalog.utils import get_programs
+from ..specializations.helpers import date_from_str
 from .helpers import get_course_open_date
```
[the remark this answers is not in the document]
Comment (monofal) [reply openedx/features/course_card/views.py:15]: I already did. These sort output has been generated by by isort that is used in LMS. Make sure you aren not using a different version because latest version tend to generate a different output

File: openedx/features/course_card/views.py [hunk 3]
```diff
@@ -1,20 +1,23 @@
...
 
 utc = pytz.UTC
 
 
 def get_course_start_date(course):
-
     """
     this function takes course and returns start date of the course
```
[the remark this answers is not in the document]
Comment (monofal) [reply openedx/features/course_card/views.py:22]: not part of my changes so not sure if I should make these changes. @saad-waqar thoughts?

File: openedx/features/course_card/views.py [hunk 4]
```diff
@@ -48,13 +50,25 @@ def get_course_cards(request):
...
+
+        # check if any active (currently open or are starting in future)
+        # run of a course is part of a program.
+        course_program = [program for program in programs
+                          for program_course in program['courses']
+                          if program_course['course_runs']
+                          and str(course.id) in [rerun['key'] for rerun in program_course['course_runs']
+                                                 if rerun['enrollment_start'] and rerun['enrollment_end']
+                                                 and (date_from_str(rerun['enrollment_start']) > current_time
+                                                 or date_from_str(rerun['enrollment_start']) <= current_time <= date_from_str(rerun['enrollment_end']))]]
```
[the remark this answers is not in the document]
Comment (monofal) [reply openedx/features/course_card/views.py:62]: this way the line is more readable. I tried breaking it down but that makes it more difficult to comprehend

File: openedx/features/course_card/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/features/course_card/views.py:22]: sorry I didn't realise it wasn't part of your changes.

File: openedx/features/course_card/views.py [hunk 5]
```diff
@@ -1,20 +1,23 @@
...
-from philu_overrides.helpers import get_user_current_enrolled_class
 from edxmako.shortcuts import render_to_response
-from openedx.features.course_card.models import CourseCard
-from django.views.decorators.csrf import csrf_exempt
 from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from openedx.features.course_card.models import CourseCard
+from philu_overrides.helpers import get_user_current_enrolled_class
 from student.models import CourseEnrollment
+
+from ...core.djangoapps.catalog.utils import get_programs
```
Comment (ahmed-faraz46) [inline openedx/features/course_card/views.py:13]: please use absolute import `from openedx.core.djangoapps.catalog.utils`

File: openedx/features/course_card/views.py [hunk 6]
```diff
@@ -1,20 +1,23 @@
...
+from ..specializations.helpers import date_from_str
```
Comment (ahmed-faraz46) [inline openedx/features/course_card/views.py:14]: please use absolute import

File: openedx/features/course_card/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahmed-faraz46) [reply openedx/features/course_card/views.py:22]: @monofal It's good to not change anything which is not part of the scope to avoid going out of the scope and increasing testing effort, but here if we have noticed something which we can improve it's not much effort and also no need to test it. So I'd also request to please change it.

File: openedx/features/course_card/views.py [hunk 7]
```diff
@@ -42,19 +44,30 @@ def get_course_cards(request):
...
+    programs = get_programs(request.site)
 
     for course in courses_list:
 
         if course.invitation_only and not CourseEnrollment.is_enrolled(request.user, course.id):
             continue
 
+        # check if any active (currently open or are starting in future)
+        # run of a course is part of a program.
+        course_program = [program for program in programs
```
Comment (ahmed-faraz46) [inline openedx/features/course_card/views.py:56]: Let's please not use list comprehension if it affects readability or where ever we need to apply it at some nested levels, I tried hard but didn't get this logic, So I would request to open it up, make it readable and not use nested comprehensions.

Comment (ahmed-faraz46) [reply openedx/features/course_card/views.py:56]: Second point is move it to a separate function and then divide that function to multiple if we can.

File: openedx/features/course_card/views.py [hunk 8]
```diff
@@ -77,6 +89,8 @@ def get_course_with_link_and_start_date(course, course_rerun_object, request):
     if current_class:
         current_class_start_date = get_course_open_date(current_class)
 
+    course.specialization = any(course_program)
```
Comment (ahmed-faraz46) [inline openedx/features/course_card/views.py:92]: Once we will remove this specialization flag from CurseOverview then we can remove this change from here. Even with current implementation instead of doing this here in this function we can do this at `71` instead of passing course_prorgram to this method.

File: openedx/features/course_card/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/features/course_card/views.py:62]: This portion of code is too complicated with high time complexity. It took me some time to understand what it is exactly doing, as pointed out by @ahmed-faraz46. Therefore, I've come up with some new implementation that divides the portion of code into multiple functions and also reduces time complexity:

First of all, such complex code should not be introduced in the existing functions, rather a new function should be created for it.

# replace existing code (line 54-62) with this
```
        if is_course_part_of_specialization(course.id, programs, current_time):
            course.specialization = True
```

# new methods

```
def is_course_part_of_specialization(course_id, programs, current_time):
    for program in programs:
        for program_course in program['courses']:
            if program_course['course_runs']:
                for rerun in program_course['course_runs']:
                    if is_course_run_active(rerun, current_time) and course_id == rerun:
                        return True
    return False


def is_course_run_active(rerun, current_time):
    if rerun['enrollment_end'] and current_time <= date_from_str(rerun['enrollment_end']):
        return True
    return False

```

[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/features/course_card/views.py:62]: Justification in favour of new implementation:

1. the new method 'is_course_run_active' more efficiently determines if a course run is active. According to previous approach, we had multiple conditionals, the need for those conditionals is eliminated in my opinion if we use this conditional only. It covers both the scenarios. If we know that the enrolment end date of a course run hasn't reached yet, it automatically implies that the course run is active. In simpler terms, the previous conditions seemed redundant.

3. the method 'is_course_part_of_specialization' is much more efficient because all we need in the output is a boolean, i.e. whether a course is part of a specialization program or not. There is no need to first get all course programs against a particular course. The moment we hit an active course run against our course id, that's it, we return true because we would add the specialisation tag to that course, if I'm not wrong. Similarly, there is no need to get all course run either, we should use the exit first approach, the moment we find what is required, break the loop/exit from the function. This will help reduce unnecessary computation of getting a list of all reruns and course programs everytime.

[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/features/course_card/views.py:62]: And @nofalfayyaz I have some other questions as well regarding the existing approach. We (@ahmed-faraz46, you and I) can discuss it in call. And @ahmed-faraz46 please correct me if there is something wrong with my suggested implementation. I have just modified the existing implementation, without changing much logic or approach. If there is altogether a different approach like you mentioned of using dictionary/not updating model, then we'll obviously go with that. But I think the existing implementation would still help in identifying courses belonging to specialisation programs.

File: openedx/core/djangoapps/content/course_overviews/models.py [hunk 9]
```diff
@@ -105,6 +105,7 @@ class Meta(object):
     eligible_for_financial_aid = BooleanField(default=True)
 
     language = TextField(null=True)
+    specialization = BooleanField(default=False)
```
[the remark this answers is not in the document]
Comment (monofal) [reply openedx/core/djangoapps/content/course_overviews/models.py:108]: @ahmed-faraz46 good point. but here we have a list of courses and we need to send this bit against each course. I thought of this initially, here what we can do is send list course id and specialization bit.
`[
    {
        course: course,
        specialization_status: [{course_id: 1, specialization},]
    }
]`
and then on the html side we would look for each course inside this list by looping through it.That basically increases the time complexity thats why I went with current approach. Let me know of your thoughts?

[the remark this answers is not in the document]
Comment (ahmed-faraz46) [reply openedx/core/djangoapps/content/course_overviews/models.py:108]: Let's please schedule a call on Monday and discuss this.

File: openedx/features/course_card/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ahmed-faraz46) [reply openedx/features/course_card/views.py:62]: Sure, Let's please schedule a call on Monday.

File: openedx/features/course_card/views.py [hunk 10]
```diff
@@ -66,8 +69,17 @@ def get_course_cards(request):
     )
 
 
-def get_course_with_link_and_start_date(course, course_rerun_object, request):
+def get_course_dict(course, programs):
+    course_dict = vars(course)
+
+    course_dict['display_name_with_default'] = course.display_name_with_default
+    course_dict['display_number_with_default'] = course.display_number_with_default
+    course_dict['specialization'] = check_course_part_of_specialization(course.id, programs)
```
Comment (zeshanarifios) [inline openedx/features/course_card/views.py:77]: Can we please change the `specialization` to `is_program_course`

File: openedx/features/course_card/views.py [hunk 11]
```diff
@@ -66,8 +69,17 @@ def get_course_cards(request):
     )
 
 
-def get_course_with_link_and_start_date(course, course_rerun_object, request):
+def get_course_dict(course, programs):
+    course_dict = vars(course)
+
+    course_dict['display_name_with_default'] = course.display_name_with_default
+    course_dict['display_number_with_default'] = course.display_number_with_default
```
[the remark this answers is not in the document]
Comment (zeshanarifios) [reply openedx/features/course_card/views.py:76]: Both are different, One is **name** and other is **number**

[the remark this answers is not in the document]
Comment (zeshanarifios) [reply openedx/features/course_card/views.py:76]: But I am just curious about why we need to set these attributes explicitly?

File: openedx/features/course_card/views.py (hunk 10 above)
Comment (zeshanarifios) [reply openedx/features/course_card/views.py:77]: I have a approach in my mind to avoid passing programs, As concerned programs will always be all programs. So instead of passing we can get programs inside the method `check_course_part_of_specialization` by using 
```
from crum import get_current_request
request = get_current_request()
programs = get_programs(request.site)
```

It ll help us to simplify the code by avoiding passing programs in get_course_dict and then check_course_part_of_specialization and decrease the number of parameters.

It would be great if we can use this.

File: openedx/features/course_card/helpers.py [hunk 12]
```diff
@@ -96,3 +97,19 @@ def get_course_cards_list():
...
+
+def check_course_part_of_specialization(course_id, programs):
+    """
+    Helper function to check if course is part of specialization
+    @param course_id: crouse key
+    @param programs: list of specialization programs
+    @return:
+    """
+    for program in programs:
+        for program_course in program['courses']:
```
[the remark this answers is not in the document]
Comment (monofal) [reply openedx/features/course_card/helpers.py:110]: yes but thats not possible as the key inside main program doesn't match with the course key.

File: openedx/features/course_card/helpers.py [hunk 13]
```diff
@@ -96,3 +97,19 @@ def get_course_cards_list():
...
+            if program_course['course_runs']:
+                for course_rerun in program_course['course_runs']:
+                    if course_rerun['key'] == str(course_id):
+                        return True
```
[the remark this answers is not in the document]
Comment (monofal) [reply openedx/features/course_card/helpers.py:114]: good suggestion. Actually early exit possible at 2nd level where we are going inside course reruns but the issue is at that level the course key that we receive from api data is different than the one that we have so that don't seem to be possible. if you have any other strategy please share.

File: openedx/features/course_card/views.py (hunk 10 above)
Comment (monofal) [reply openedx/features/course_card/views.py:77]: @zeshanarifios actually that function is getting called inside a loop. if we take `get_programs` inside `check_course_part_of_specialization` then it will retrieve data from cache against each course. according to current approach it retrieve it only once.

File: openedx/features/course_card/views.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (monofal) [reply openedx/features/course_card/views.py]: @harris-irfan @zeshanarifios `display_name_with_default` is model method. previously we were sending sending objects to front end page but now we are sending a dictionary and we can't call model method. thats why I have to send it over to front end in a key. `display_number_with_default` has been removed

File: openedx/features/course_card/helpers.py [hunk 14]
```diff
@@ -96,3 +97,19 @@ def get_course_cards_list():
     course_card_ids = [cc.course_id for cc in cards_query_set]
     courses_list = CourseOverview.objects.select_related('image_set').filter(id__in=course_card_ids)
     return courses_list
+
+
+def check_course_part_of_specialization(course_id, programs):
```
[the remark this answers is not in the document]
Comment (harris-irfan) [reply openedx/features/course_card/helpers.py]: 'is_course_in_programs' seems ok to me. It's just that for the scope of the current ticket, we are adding 'Specialisation' tags. That's why I chose the word 'specialisation'. However, for consistency reasons, we can use 'programs'.

File: openedx/features/course_card/views.py (hunk 10 above)
Comment (zeshanarifios) [reply openedx/features/course_card/views.py]: @monofal 👍 
How about calling it inside `get_course_dict`, It ll decrease 1 level and if it needs to be called anywhere else we don't need to pass it again.

File: openedx/features/course_card/helpers.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zeshanarifios) [reply openedx/features/course_card/helpers.py:114]: @monofal I didn't get your point that why it isn't possible. 
Is code(with early exit) I shared above will not work for your case? 
I added early exit at this line by inverting your condition, 
```
if not program_course['course_runs']:
      continue
```

File: openedx/features/course_card/views.py [hunk 15]
```diff
@@ -66,8 +69,16 @@ def get_course_cards(request):
     )
 
 
-def get_course_with_link_and_start_date(course, course_rerun_object, request):
+def get_course_dict(course, programs):
```
Comment (ahmed-faraz46) [inline openedx/features/course_card/views.py:72]: How about to move this method in helpers?

Comment (monofal) [reply openedx/features/course_card/views.py:72]: its for course data population and gives code reader a better readability. Moreover, I don't think it falls under the category of helpers because the implementation is explicitly for this specific view. thoughts?

Comment (ahmed-faraz46) [reply openedx/features/course_card/views.py:72]: I didn't get your readability point, Actually this is a method which takes a course and a program and tells either this course is part of the program or not it's being used in view but not a view.
Let's hop on a call if still there is any confusion.

**Comment by monofal** [comment]
Re opening this PR as it was merged to master instead of develop.
